Issue 387  spect

niinimaki opened the pull request:
Add SPECT projectors using NiftyRec as back-end.
Also a test for `spect_trafo.py` is added.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        self._attenuation = attenuation
+        self._psf = psf
+
+        lowleft_x = geometry.detector.params.corners()[0][0]
```
`[lowleft_x, lowleft_y], [upright_x, upright_y] = geometry.detector.params.corners()`

niinimaki replied:
Except that this `[lowleft_x, lowleft_y], [upright_x, upright_y] = geometry.detector.params.corners()` did not work.
But I'll do this more economically as @kohr-h  wrote above.

adler-j replied:
It should. What error do you get?

adler-j replied:
Ah, it seems you need:

```python
[lowleft_x, lowleft_y], _, _, [upright_x, upright_y] = geometry.detector.params.corners()
```

Another option is to do:

```python
[lowleft_x, lowleft_y] = geometry.detector.params.min()
[upright_x, upright_y] = geometry.detector.params.max()
```

niinimaki replied:
`[lowleft_x, lowleft_y], [upright_x, upright_y] = geometry.detector.params.corners()`
`ValueError: too many values to unpack`

But this works

`geom_corners = geometry.detector.params.corners()`
` [lowleft_x, lowleft_y] = geom_corners[0]`
` [upright_x, upright_y] = geom_corners[-1]`

and is more economical than calling `geometry.detector.params.corners()` four times.

kohr-h replied:
You can unpack like this:
```python
ll, ul, lr, ur = space.grid.corners()
```
However, my impression is that this whole section is unnecessary.

niinimaki replied:
> EDIT: You need to put the angles last, right? Then, do ...

Yes. I'll do the modifications.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        lowleft_x = geometry.detector.params.corners()[0][0]
+        lowleft_y = geometry.detector.params.corners()[0][1]
+        upright_x = geometry.detector.params.corners()[-1][0]
+        upright_y = geometry.detector.params.corners()[-1][1]
```
Why -1 in first and 1 in second?

niinimaki replied:
To get the second element from the last element of `geometry.detector.params.corners()`, but your comment above is much nicer way of performing this.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        self.geometry = geometry
+        det_nx = geometry.det_grid.shape[0]
+        det_ny = geometry.det_grid.shape[1]
+        startAngle = float(geometry.motion_grid.points()[0])
```
`startAngle = geometry.angles[0]`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        det_nx = geometry.det_grid.shape[0]
+        det_ny = geometry.det_grid.shape[1]
+        startAngle = float(geometry.motion_grid.points()[0])
+        stopAngle = float(geometry.motion_grid.points()[-1]) * (180 / np.pi)
```
Why convert this to degrees but not start? Also, we usually try to avoid degrees everywhere just to keep consistent.

niinimaki replied:
I had been thinking that start is 0.0, which it, of course, is not necessarily. I'll remove the conversion to degrees.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        out : `DiscreteLpVector`
+            Returns an element in the projection space
+        """
+        if self._attenuation is None:
```
Any reason this is not done in the initializer?

niinimaki replied:
Because of the following `else` statement and afterwards this "no-longer ODL space element" is passed to the `SpectBackproject` which requires `attenuation` to be a `element in :class: ~odl.discr.lp_discr.DiscreteLp`

adler-j replied:
Lets leave it for now, but i think we can optimize this later if we find that it takes time.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        projection = SPECT_project_parallelholes(x, self._cameras,
+                                                 attenuation, psf,
+                                                 use_gpu=self._use_gpu)
+        return self.range.element(projection)
```
`return projection`

Cast is implicit

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        psf : `DiscreteLpVector`
+            Point Spread Functions for the correction of the
+            collimator-detector response
+        use_gpu : `bool`, optional
```
use_gpu should be part of the backend, similar to how we do it with astra. I.e.

`impl=niftyrec_gpu`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+from odl.operator.operator import Operator
+
+
+_SUPPORTED_IMPL = ('niftyrec')
```
This should come after `__all__`

kohr-h replied:
And be a tuple, i.e. `('niftyrec',)`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+import numpy as np
+
+# Internal
+import odl
```
This will cause import problems

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+
+# Internal
+import odl
+from odl.operator.operator import Operator
```
`from odl.operator import Operator`

should suffice

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        psf : element in :class:`~odl.discr.lp_discr.DiscreteLp`
+            Point Spread Functions  for the correction of the
+            collimator-detector response
+        use_gpu : `bool`, optional
```
See above comment on use_gpu

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        ----------
+        dom : `DiscreteLp`
+            Discretized space, the domain of the forward projector
+        geometry : `Geometry`
```
No harder requirements than `Geometry`? Perhaps `SpectGeometry` or similar?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+            The geometry of the transform. An ODL geometry instance
+            that contains information about the operator domain
+        attenuation : `DiscreteLpVector`
+            Linear attenuation map for SPECT.
```
Element in what space?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        attenuation : `DiscreteLpVector`
+            Linear attenuation map for SPECT.
+        psf : `DiscreteLpVector`
+            Point Spread Functions for the correction of the
```
Element in what space?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        Parameters
+        ----------
+        dom : `DiscreteLp`
+            Discretized space, the domain of the forward projector
```
Requirements on the space? Dimension, storage?

niinimaki replied:
> Requirements on the space?

I don't quite see what you mean by that...

adler-j replied:
That this method has requirements on the space. It cannot be complex valued. It cannot be stored in cuda, it has to be 3 dimensional (or does it handle 2?). Perhaps there are more.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+class SpectProject(Operator):
+    """Spect projector using backend NiftyRec"""
+    def __init__(self, dom, geometry, attenuation, psf, use_gpu=False,
+                 impl='niftyrec', **kwargs):
```
**kwargs not needed since its not used

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+class SpectBackproject(Operator):
+
+    def __init__(self, ran, geometry, attenuation, psf, use_gpu=False,
+                 impl='niftyrec', **kwargs):
```
**kwargs not needed since its not used

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        activity = SPECT_backproject_parallelholes(x, self._cameras,
+                                                   attenuation, psf,
+                                                   use_gpu=self._use_gpu)
+        return self.range.element(activity)
```
return activity

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+                             ''.format(impl_in))
+
+        self.geometry = geometry
+        det_nx = geometry.det_grid.shape[0]
```
det_nx, det_ny = geometry.det_grid.shape

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+            raise ValueError('implementation {!r} not supported.'
+                             ''.format(impl_in))
+
+        self.geometry = geometry
```
Why is geometry public but the others not?

niinimaki replied:
No strong reason.

kohr-h replied:
We have been a bit on the cautious side regarding public attributes so far. In general, all attributes are private first (i.e. start with an underscore) until there is a good reason to expose them, for example they are very useful in application code or they are used in other parts of the code outside the class.
I'd say `geometry` and `attenuation` are interesting enough to the outside world, `psf` probably not. That may change.

So, as a general guideline, *always* use the underscored version
```python
self._attribute = value
```
and add
```python
@property
def attribute(self):
    """Description of the attribute."""
    return self._attribute
```
if you think the attribute is needed in outside code.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        det_ny = geometry.det_grid.shape[1]
+        startAngle = float(geometry.motion_grid.points()[0])
+        stopAngle = float(geometry.motion_grid.points()[-1]) * (180 / np.pi)
+        N_proj = len(geometry.motion_grid.points())
```
`N_proj = geometry.motion_grid.size`

kohr-h replied:
Another general thing: If you call a method which is not a property, i.e. you use parentheses like for `grid.points()`, you need to assume that it comes at a computational cost. So calling it three times just to pick out one value adds unnecessary overhead (not so much for 1d angles, but anyway).
If you see yourself using the result of such a method call more than once, it may be a good idea to store the result. See also the use of `corners` below.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+standard_library.install_aliases()
+from builtins import super
+
+# External
```
We removed these external, internal comments. Just keep ODL imports below and grouped.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+__all__ = ('SpectProject', 'SpectBackproject')
+
+
+class SpectProject(Operator):
```
If we can make sure that it's (roughly) true mathematically, we should call this guy `AttenuatedRayTransform`.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+
+
+class SpectProject(Operator):
+    """Spect projector using backend NiftyRec"""
```
Currently Nifty is our only back-end, but still we should not name it explicitly here. Rather, the class docstring should contain a short explanation about the nature of the operator (not using any LaTeX).

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        geometry : `Geometry`
+            The geometry of the transform. An ODL geometry instance
+            that contains information about the operator domain
+        attenuation : `DiscreteLpVector`
```
We should accept anything that can be *converted* into an adequate space element. The adequate space could be the same as this operator's domain or some other space with the same domain of definition.
The simplest assumption would be the former, which would then simply mean that [this line](https://github.com/odlgroup/odl/pull/391/files#diff-f72b925327f97605e512656ea6a10472R85) is replaced with 
```python
self._attenuation = self.domain.element(attenuation)
```
Support for finer or coarser spaces can be added later. Should be simple with the `Resampling` operator, however. Or even doable by hand.

niinimaki replied:
Do you mean that we would only require `attenuation` (and `psf`) to be a numpy arrays?

kohr-h replied:
Not even that. I mean [element-like](https://odl.readthedocs.io/guide/glossary.html#term-element-like), which in this case means anything that the subsequent call `self.domain.element()` accepts. That could already be a space element, but also a Numpy array or even a Python function.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        use_gpu : `bool`, optional
+            'False' if not using GPU
+            `True` if using GPU provided that `impl` supports it
+        impl='niftyrec', optional
```
Parameters with fixed possible values are written as
```
impl : {'niftyrec'}, optional
    Implementation back-end for the transform. Supported back-ends:

    'niftyrec' : NiftyRec tomography toolbox using the niftypy Python wrapper
```
A link to the NiftyRec (or NiftyPy if it's good) home page would also be good to have.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        upright_y = geometry.detector.params.corners()[-1][1]
+        ran = odl.uniform_discr([lowleft_x, lowleft_y, startAngle],
+                                [upright_x, upright_y, stopAngle],
+                                [det_nx, det_ny, N_proj])
```
All this code can basically be compressed into one line:
```python
ran = uniform_discr_frompartition(geometry.partition)
```

adler-j replied:
Agree. This is the change we need.

kohr-h replied:
Copy & paste of the edit above:

EDIT: You need to put the angles last, right? Then, do 
```
partition = geometry.det_partition.append(geometry.motion_partition)
ran = uniform_discr_frompartition(partition)
```

adler-j replied:
Well, better would be if this had the angles first, as that would be expected from the geometry.

adler-j replied:
Regarding the above. I'm even more sure in my conviction here now that I realize that this is how RayTransform does it. We should strive for similarity here.

In ray transform we swap the axes of the input (which costs us some time) in order to allow a uniform interface. This makes it easier to extend with more backends.

kohr-h replied:
Actually, you're right. No reason why the angles should come last.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+            that contains information about the operator domain
+        attenuation : `DiscreteLpVector`
+            Linear attenuation map for SPECT.
+        psf : `DiscreteLpVector`
```
optional

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+
+class SpectProject(Operator):
+    """Spect projector using backend NiftyRec"""
+    def __init__(self, dom, geometry, attenuation, psf, use_gpu=False,
```
`psf` should be optional

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        N_proj = len(geometry.motion_grid.points())
+
+        self._cameras = np.float32(np.linspace(startAngle, stopAngle, N_proj,
+                                   dtype=np.float32).reshape((N_proj, 1)))
```
Double cast to `float32`

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+                                use_gpu=self._use_gpu)
+
+
+class SpectBackproject(Operator):
```
The initializer of this guy is very much copy & paste from the forward operator. I've found the following pattern to be very useful in that case:

- Make the adjoint a subclass of the forward operator
- In the `__init__` function, run the parent init first, then switch domain and range.
- Override `_call` and `adjoint`.

Saves about 80 % of duplicated code.

I can help out if needed.

adler-j replied:
* Make the adjoint a subclass of the forward operator

Disagree here. Inheritance should follow **is a**, but the `SpectBackproject` is obviously not a `SpectProject`

kohr-h replied:
That's not the pythonic view.

adler-j replied:
Source on that? I feel it would be very wierd. It would for example break any `isinstance(SpectProject)` since it doesnt mean anything

kohr-h replied:
[Here](https://en.wikipedia.org/wiki/Duck_typing)'s a good source.

Regarding the issue identifying a suitable SPECT projector: why do we ever want to do that? If code depends on a combination of a projector and data, the error message when trying to apply the adjoint of the projector to the data will be quite obvious.

adler-j replied:
Dont really see any mention of inheritance there. For me the above design would be akin to doing (with reference to the python example)

```python
class Duck:
    def __init__(self, name):
        self.name = name
        
    def quack(self):
        print("Quack, quack!")


class Person(Duck):
    def quack(self):
        print("I'm Quackin'!")
```

Sure, it makes sense from a programming point, but it is also ambiguous. Why not change `Duck` to inherit from `Person`? What happens if I want to add properties to `Duck`, do I need to `del` them in `Person`? The "obvious" design would be to have a common point of inheritance

```python
class Named:
    def __init__(self, name):
        self.name = name

        
class Duck(Named):        
    def quack(self):
        print("Quack, quack!")

        
class Person(Named):
    def quack(self):
        print("I'm Quackin'!")
```

This would keep the dependencies separated and wouldnt feel as "arbitrary". 

I remain highly sceptical. Have you ever seen this type of design used in another library so I can study an example of it?

kohr-h replied:
I certainly see your point. That's relevant as soon as we need something in the parent that absolutely shouldn't be in the subclass:
```python
class Duck:
    def __init__(self, name):
        self.name = name
        
    def quack(self):
        print("Quack, quack!")

    def fly(self, height):
        self.height = height

class Person(Duck):
    def quack(self):
        print("I'm Quackin'!")

    def fly(self, height):
        raise NotImplementedError('I cannot fly!')
```
Currently our guys are similar enough not to come into this situation, but okay, it cannot be excluded. Better to have a common parent class then to save code.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,226 @@
+        lowleft_y = geometry.detector.params.corners()[0][1]
+        upright_x = geometry.detector.params.corners()[-1][0]
+        upright_y = geometry.detector.params.corners()[-1][1]
+        dom = odl.uniform_discr([lowleft_x, lowleft_y, startAngle],
```
See above discussion on how to simplify this

niinimaki commented:
Improve the calling and docs of `spect_trafo.py` as follows (according to reviews):

- improved imports

- `attenuation` and `psf` are accepted as element-like and are optional

- improved docstrings, for example added links to NiftyRec and NiftyPy

- simplified initializations of `range` and `domain` in `SpectProject` and `SpectBackproject`, respectively

- Angles are now first in `range` of `SpectProject`

- made the selection between CPU and GPU to be part of the backend

At least remains to do:

> Currently Nifty is our only back-end, but still we should not name it explicitly here. Rather, the class docstring should contain a short explanation about the nature of the operator (not using any LaTeX).

at this point only removed the name "NiftyRec"  from the description.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+wordpress/>`_ tomography toolbox using the `niftypy <https://github.com/\
+spedemon/niftypy>`_ Python wrapper and CPU
+
+            'niftyrec_gpu' : `NiftyRec <http://niftyrec.scienceontheweb.net/\
```
No need to duplicate link

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        psf : `element-like` structure, optional
+            Point Spread Functions for the correction of the
+            collimator-detector response
+        impl={'niftyrec', 'niftyrec_gpu'}, optional
```
replace "=" by " : " (with the spaces)

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, x):
+        """Compute the discrete
```
More text I suppose

niinimaki replied:
:) same as above

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+_SUPPORTED_IMPL = ('niftyrec', 'niftyrec_gpu')
+
+
+class SpectProject(Operator):
```
Since Project is a word on its own, Id go for the full `SpectProjector` or perhaps `AttenuatedRayTransform` as suggested by @kohr-h.

kohr-h replied:
`AttenuatedRayTransform`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        det_nx, det_ny = geometry.det_grid.shape
+        startAngle = geometry.angles[0]
+        stopAngle = geometry.angles[-1]
+        N_proj = geometry.motion_grid.size
```
`N_proj = geometry.angles.size`

would be even shorter

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+
+            'niftyrec_gpu' : `NiftyRec <http://niftyrec.scienceontheweb.net/\
+wordpress/>`_ tomography toolbox using the `NiftyPy <https://github.com/\
+spedemon/niftypy>`_ Python wrapper and GPU
```
This looks a bit ugly in the parameter list. I suggest making a `References` section by adding
```
References
----------
.. _NiftyRec: http://niftyrec.scienceontheweb.net/wordpress/
.. _NiftyPy: https://github.com/spedemon/niftypy
```
and referring in the text by
```
`NiftyRec`_
```
and
```
`NiftyPy`_
```
respectively.

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,72 @@
+    vol = domain.one()
+    vol = odl.util.phantom.shepp_logan(domain, True)
+
+    if NIFTYREC_AVAILABLE:
```
This should come above with some kind of skip statement, see the `RayTransform` tests for examples

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,72 @@
+
+        assert proj in projector.range
+        assert backproj in projector.domain
+
```
Verify that the adjoint property holds

adler-j replied:
Perhaps test some well known property of this transform, maybe use some very simple data with a known result?

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,72 @@
+    # Create a discrete domain and a phantom as an element of the domain
+    domain = odl.uniform_discr([-20, -20, -20], [20, 20, 20],
+                               [det_nx_pix, det_nx_pix, det_nx_pix])
+    vol = domain.one()
```
redundant

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,72 @@
+                                              det_rad=det_radius)
+
+    # Create a discrete domain and a phantom as an element of the domain
+    domain = odl.uniform_discr([-20, -20, -20], [20, 20, 20],
```
`domain = odl.uniform_discr([-20] * 3, [20] * 3, [det_nx_pix] * 3)

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,72 @@
+
+# Internal
+import odl
+from odl.tomo.geometry.spect import ParallelHoleCollimatorGeometry
```
Simply use the path inside ODL, remove these two lines

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        N_proj = geometry.motion_grid.size
+
+        self._cameras = np.linspace(startAngle, stopAngle, N_proj,
+                                    dtype=np.float32).reshape((N_proj, 1))
```
`self._cameras = geometry.angles[:, None]` should work.

kohr-h replied:
I think this is almost too trivial to store. Also, it's specific to the Nifty back-end, so I'd say use the `geometry.angles[:, None]` object directly in `_call` and don't store it here.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        det_nx, det_ny = geometry.det_grid.shape
+        startAngle = geometry.angles[0]
+        stopAngle = geometry.angles[-1]
+        N_proj = geometry.motion_grid.size
```
No capitals in variable names. But it's not needed anyway.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+                self._use_gpu = False
+
+        self._geometry = geometry
+        det_nx, det_ny = geometry.det_grid.shape
```
unused?

niinimaki replied:
True (after the simplifications, that is)

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        self._cameras = np.linspace(startAngle, stopAngle, N_proj,
+                                    dtype=np.float32).reshape((N_proj, 1))
+        if psf is not None:
+            self._psf = np.asarray(psf)
```
Why is psf converted using asarray while attenuation is converted using self.domain.element?

kohr-h replied:
@niinimaki and I discussed this. Usually the PSF is given as an array of small size and cannot simply be made an element of the domain. And I assume (without knowing better) that blowing it up to full size would bog down the code significantly. Would be worth checking, though. This is the obvious fix, delegating the PSF handling to Nifty.

If there is no difference in speed, we could simply add zeros to the kernel and go on. Otherwise, I'd add a TODO here, something like
```
# TODO: Create a smaller space for the PSF if necessary and convert to an element.
```

adler-j replied:
Well unless there is an agreed uppon definition of the psf (which, imo would be a element in the range or similar) then it should be moved to **kwargs and `Other Parameters`. That would make this "hacky" handling of it more ok. Also  perhaps a note that it is Nifty dependent.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+                raise NotImplementedError('NiftyRec GPU back-end only '
+                                          'supports volume size '
+                                          '(128, 128, 128), got {}'
+                                          ''.format(dom.shape))
```
Perfect!

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, x):
+        """Compute the discrete SPECT ptojection
```
Dot at the end. Also typo "ptojection".

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+
+
+class SpectProject(Operator):
+    """Discrete SPECT projector"""
```
Class docstrings have blank lines above and below.

adler-j commented:
Certainly taking shape, mostly down to style now.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+
+    @property
+    def adjoint(self):
+        """The back-projector associated with this array."""
```
This array? Operator.

niinimaki replied:
Oops a bigger type here :)

niinimaki replied:
> type

I meant typo...

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        Parameters
+        ----------
+        x : `DiscreteLpVector`
+           Element in the domain of the operator to be forward projected
```
Very boilerplate, perhaps "The volume that should be forward projected" or something like that.

niinimaki replied:
This was actually just a copy-paste from [`ray_trafo`](https://github.com/odlgroup/odl/blob/master/odl/tomo/operators/ray_trafo.py#L169) :)

adler-j replied:
*wistles* 

Perhaps that needs a update as well then :)

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        Returns
+        -------
+        out : `DiscreteLpVector`
+            Returns an element in the projection space
```
See above, some description on what it does would be nice.

niinimaki replied:
as above, same [here](https://github.com/odlgroup/odl/blob/master/odl/tomo/operators/ray_trafo.py#L177), but I'll try to be more informative

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+
+        Returns
+        -------
+        out : `DiscreteLpVector`
```
At this point we usually write

`out : \`\`range\`\` element`

This could be applied to other places as well

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+                                          ''.format(dom.shape))
+            if impl == 'niftyrec_gpu':
+                self._use_gpu = True
+            elif impl == 'niftyrec':
```
else:

adler-j replied:
Honestly this can be written:

```python
self._use_gpu = impl == 'niftyrec_gpu'
```

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        ----------
+        dom : `DiscreteLp`
+            Discretized space, a 3 dimensional domain of the forward projector
+        geometry : geometry from :class:`~odl.tomo.geometry.spect`
```
geometry : `ParallelHoleCollimatorGeometry`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        geometry : geometry from :class:`~odl.tomo.geometry.spect`
+            The geometry of the transform. An ODL geometry instance
+            that contains information about the operator domain
+        attenuation : `element-like` structure of the shape
```
```
attenuation : domain `element-like`
```

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        attenuation : `element-like` structure of the shape
+            `dom.shape`, optional
+            Linear attenuation map for SPECT.
+        psf : `element-like` structure, optional
```
```
psf : domain `element-like`
```

But note discussion below. Is this really the case? Otherwise more information is needed.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+                self._use_gpu = True
+            elif impl == 'niftyrec':
+                self._use_gpu = False
+
```
The usual `isinstance(geometry, CLASS NAME)` goes here

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        ran = uniform_discr_frompartition(geometry.partition)
+        super().__init__(dom, ran, linear=True)
+
+    def _call(self, x):
```
Currently, the Nifty back-end is hard-coded. To make this more extendible, we should rename the current `_call` to `_call_nifty` and write a new `_call` which just decides which one of the `_call_*` to take. See [here](https://github.com/odlgroup/odl/blob/master/odl/trafos/fourier.py#L1592-L1617) for an example. The main documentation goes into the `_call` method, the other guys don't need so much documentation. Check [here](https://github.com/odlgroup/odl/blob/master/odl/trafos/fourier.py#L1660-L1672) for an example.

If we get more back-ends, we will have to factor out those specific parts into own files like for the ray transform, but for now it's okay like this.

Note: This requires us to store `impl`. Would be worth a property also.

kohr-h commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,254 @@
+        self._impl = impl
+
+        ran = uniform_discr_frompartition(geometry.partition)
+        super().__init__(dom, ran, linear=True)
```
This call must come much earlier. Before this, any reference to `self.domain` or `self.range` will fail.

Apart from that, it should be checked that `geometry` is a reasonable (supported) geometry (perhaps add another list for those), and that the domain is an adequate space. I guess the latter means `DiscreteLp` and some more things.

niinimaki replied:
Sure thing, I'll add the checks.

niinimaki commented:
According to the review:

- improved docstrings (added links to `NiftyRec` and `NiftyPy` to the `References`, for example)

- changed the name `SpectProject` -> `AttenuatedRayTransform` according to @kohr-h  suggestion

- added `_call_niftyrec` 

- removed redundant variables

Remains (at least): improve `spect_trafo_test.py`

niinimaki commented:
Related to `spect_trafo`:

- Added an adjoint test for the `spect_trafo`

- Improved the docs a bit more (corrected some typos etc).

Review?

niinimaki commented:
Could I have a review for this one?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+
+class AttenuatedRayTransform(Operator):
+
+    """The discrete attenuated Ray transform between L^p spaces."""
```
Perhaps a word on what this actually means. What does it compute?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        geometry : `ParallelHoleCollimatorGeometry`
+            The geometry of the transform. An ODL geometry instance
+            that contains information about the operator range
+        attenuation : domain `element-like` structure of the shape
```
attenuation : domain `element-like`

should do

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        References
+        ----------
+        .. _NiftyRec: http://niftyrec.scienceontheweb.net/wordpress/
+
```
No blank row

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+
+        """
+        if not isinstance(dom, DiscreteLp):
+            raise TypeError('discretized domain {!r} is not a `DiscreteLp`'
```
~~discretized~~

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+            if impl == 'niftyrec_gpu' and dom.shape != (128, 128, 128):
+                raise NotImplementedError('NiftyRec GPU back-end only '
+                                          'supports volume size '
+                                          '(128, 128, 128), got {}'
```
dot at the end

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        Parameters
+        ----------
+        dom : `DiscreteLp`
+            Discretized space, a 3 dimensional domain of the forward projector
```
Domain of the forward projector, 3 dimensional.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        dom : `DiscreteLp`
+            Discretized space, a 3 dimensional domain of the forward projector
+        geometry : `ParallelHoleCollimatorGeometry`
+            The geometry of the transform. An ODL geometry instance
```
Lots of redundancy here. "Geometry of the transform" should do, unless we need to add more information.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        attenuation : domain `element-like` structure of the shape
+            `dom.shape`, optional
+            Linear attenuation map for SPECT.
+        psf : `element-like` structure, optional
```
element in what?

Also, remove structure

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+                                          '(128, 128, 128), got {}'
+                                          ''.format(dom.shape))
+            if impl == 'niftyrec_gpu':
+                self._use_gpu = (impl == 'niftyrec_gpu')
```
Change this if ... else ... statement to

`self._use_gpu = (impl == 'niftyrec_gpu')`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        if self._attenuation is None:
+            attenuation = None
+        else:
+            attenuation = np.float32(
```
`attenuation  = np.asfortranarray(self._attenuation, dtype='float32')`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        else:
+            psf = np.float32(self._psf)
+
+        x = np.float32(np.asfortranarray(x))
```
x = np.asfortranarray(x, dtype='float32')

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        if self._psf is None:
+            psf = None
+        else:
+            psf = np.float32(self._psf)
```
How come psf does not need to be a fortranarray?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+from builtins import super
+
+try:
+    from NiftyPy.NiftyRec import (SPECT_project_parallelholes,
```
It's probably better to keep the namespace, makes it more obvious what is an external library call:

`from NiftyPy import NiftyRec`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+                             ''.format(impl_in))
+
+        if impl.startswith('niftyrec'):
+            if not NIFTYREC_AVAILABLE:
```
Is `niftyrec_gpu` always available if `niftyrec` is?

niinimaki replied:
Yes it is, the only thing here is that in the current version of NiftyRec,  `niftyrec_gpu` performs consistently only with volume size `128 x 128 x 128`.

adler-j replied:
So if I have a computer without a GPU, I can still use `niftyrec_gpu`? What happens in that case?

niinimaki replied:
Yes sure of course, if one does not have CUDA... 
Actually I don't know what happens, maybe an error message from NiftyRec...I'll try to figure that out.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        References
+        ----------
+        .. _NiftyRec: http://niftyrec.scienceontheweb.net/wordpress/
+
```
blank row

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+        ran : `DiscreteLp`
+            Reconstruction space, the range of the back-projector
+        geometry : `ParallelHoleCollimatorGeometry`
+            The geometry of the transform. An ODL geometry instance
```
Many of the geometry comments from above apply here as well.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,331 @@
+
+class AttenuatedRayBackprojection(Operator):
+
+    """The adjoint of the discrete attenuated Ray transform
```
This should be a one line description.

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,78 @@
+                                  [det_nx_pix, det_ny_pix])
+
+    apart = odl.uniform_partition(0, 2 * np.pi, n_proj)
+    geometry = odl.tomo.geometry.spect.ParallelHoleCollimatorGeometry(
```
`odl.tomo.geometry.ParallelHoleCollimatorGeometry` should be sufficient, perhaps add `spect` to `geometry.__init__`

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,78 @@
+
+
+@skip_if_no_niftyrec
+def test_spect_projector():
```
This should be expanded:

* Test gpu
* Test attenuation
* Test PSF
* Test that the calculated values are correct in some sense (i.e. the correct "shape" according to angles and the correct "magnitude" according to our definition of the attenuated ray transform).

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        References
+        ----------
+        .._NiftyRec: http://niftyrec.scienceontheweb.net/wordpress/
+
```
Remove blank row

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        .._NiftyRec: http://niftyrec.scienceontheweb.net/wordpress/
+
+        .._NiftyPy: https://github.com/spedemon/niftypy
+
```
Remove blank row

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        impl : {'niftyrec', 'niftyrec_gpu'}, optional
+            Implementation back-end for the transform. Supported back-ends:
+
+            'niftyrec' : `NiftyRec`_ tomography toolbox using the `NiftyPy`_
```
Notation is 'astra_cpu'/'astra_cuda' for astra, perhaps we should decide on a style and use the same here.

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+
+        if not isinstance(geometry, ParallelHoleCollimatorGeometry):
+            raise TypeError('geometry {!r} is not a'
+                            '`ParallelHoleCollimatorGeometry` instance.'
```
no dot at end

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+
+        impl, impl_in = str(impl).lower(), impl
+        if impl not in _SUPPORTED_IMPL:
+            raise ValueError('implementation {!r} not supported.'
```
no dot at end

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+            if not NIFTYREC_AVAILABLE:
+                raise ValueError('NiftyRec back-end not available.')
+            if impl == 'niftyrec_gpu' and dom.shape != (128, 128, 128):
+                raise NotImplementedError('`nifryrec_gpu` back-end only '
```
Use single ticks to show that niftyrec_gpu is a string. Also spelling error.

"'niftyrec_gpu' back-end only "

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        if self._impl.startswith('niftyrec'):
+            return self._call_niftyrec(x.asarray())
+        else:
+            raise NotImplementedError('Only NiftyRec back-end supported '
```
only

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        x :  `numpy.ndarray`
+           The volume that is forward projected. The volume is an element
+           in the domain of the operator.
+
```
Remove blank row

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+                                          '(128, 128, 128), got {}'
+                                          ''.format(ran.shape))
+            self._use_gpu = (impl == 'niftyrec_gpu')
+
```
else:
    raise ValueError

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+                                          '(128, 128, 128), got {}'
+                                          ''.format(dom.shape))
+            self._use_gpu = (impl == 'niftyrec_gpu')
+
```
```
else:
    raise ValueError(...)
```

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+        cameras = np.float32(self._geometry.angles[:, None])
+        activity = NiftyRec.SPECT_backproject_parallelholes(
+            x, cameras, attenuation, psf, use_gpu=self._use_gpu)
+        scale = self._geometry.det_partition.cell_sides.prod()
```
`scale = self._geometry.det_partition.cell_volume`

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+            x, cameras, attenuation, psf, use_gpu=self._use_gpu)
+        scale = self._geometry.det_partition.cell_sides.prod()
+        scale *= self._geometry.motion_partition.cell_volume
+        scale /= self.range.partition.cell_sides.prod()
```
`scale /= self.range.partition.cell_volume`

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+# Internal
+import odl
+from odl.tomo.operators.spect_trafo import AttenuatedRayTransform
+from odl.util.testutils import skip_if_no_niftyrec
```
```
from odl.util.testutils import (skip_if_no_niftyrec,
                                almost_equal, all_equal)
```

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+    proj_no_att = projector(vol_element)
+    # Accept 0.1 % error
+    places = 3
+    assert odl.util.testutils.almost_equal(np.max(proj_no_att), 100,
```
`assert almost_equal(`

adler-j replied:
Also it feels like this is not the correct value. Shouldnt the maximum scale with the voxel size?

niinimaki replied:
Seems that it is not scaled in `NiftyRec`..

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+    proj_att = projector_att(vol_element)
+
+    value = np.sum(vol * np.exp(-np.sum(attenuation)))
+    assert odl.util.testutils.almost_equal(np.max(proj_att), value,
```
`assert almost_equal(`

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+
+    assert all_equal(proj.shape, (n_proj, det_nx_pix, det_ny_pix))
+    assert all_equal(backproj.shape, vol.shape)
+    assert proj in projector.range
```
this really aint needed, operators already do this check

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+
+        Returns
+        -------
+        out : element in the ``domain`` of the operator
```
Quite sure `out` is an `numpy.ndarray` here?

adler-j commented on odl/tomo/operators/spect_trafo.py:
```diff
@@ -0,0 +1,318 @@
+
+        Returns
+        -------
+        out : ``range`` element
```
`out` is an `numpy.ndarray` in code

niinimaki replied:
Doesn't [this](https://github.com/odlgroup/odl/blob/issue-387__spect/odl/tomo/operators/spect_trafo.py#L172) automatically return `out` as an element of `range`...

adler-j replied:
ODL doesnt know about `_call_niftyrec` so `_call_niftyrec` still returns whatever it wants. The magic only happens if you call `__call__` which in turn does its magic on the return value of `_call`.

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+        domain, geometry, attenuation=attenuation, impl='niftyrec')
+    proj_att = projector_att(vol_element)
+
+    value = np.sum(vol * np.exp(-np.sum(attenuation)))
```
I'm not quite sure I follow. Why is this the analytic answer?

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+    # Create a discrete domain and a phantom as an element of the domain
+    domain = odl.uniform_discr([-1] * 3, [1] * 3, [det_nx_pix] * 3)
+
+    vol = np.zeros((16, 16, 16))
```
`vol = np.zeros([det_nx_pix] * 3)`

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+
+    apart = odl.uniform_partition(0, 2 * np.pi, n_proj)
+    geometry = odl.tomo.geometry.spect.ParallelHoleCollimatorGeometry(
+        apart, dpart, det_rad=det_radius)
```
What is `det_radius` used for in the computation?

niinimaki replied:
In [spect geometry](https://github.com/odlgroup/odl/blob/issue-387__spect/odl/tomo/geometry/spect.py#L71). But not anywhere else, since unfortunately `NiftyRec` does not support contoured orbits..

adler-j replied:
Not quite sure why it's needed in that case. Shouldnt `AttenuatedRayTransform` work with a `Parallel3dAxisGeometry` then?

niinimaki replied:
I suppose it should..

adler-j commented on test/tomo/operators/spect_trafo_test.py:
```diff
@@ -0,0 +1,123 @@
+    value = np.sum(vol * np.exp(-np.sum(attenuation)))
+    assert odl.util.testutils.almost_equal(np.max(proj_att), value,
+                                           places=places)
+
```
Another note for these tests. Should'nt we expect the following to hold:

```
det_nx_pix = 16
det_ny_pix = 16
det_nx_mm = 0.4
det_radius = 20
n_proj = 1
det_param = det_nx_mm * det_nx_pix

domain = odl.uniform_discr([-1] * 3, [1] * 3, [det_nx_pix] * 3)

dpart = odl.uniform_partition([-det_param, -det_param],
                              [det_param, det_param],
                              [det_nx_pix, det_ny_pix])

apart = odl.uniform_partition(0, 2 * np.pi, n_proj)

geometry1 = odl.tomo.geometry.spect.ParallelHoleCollimatorGeometry(
    apart, dpart, det_rad=det_radius)
geometry2 = odl.tomo.geometry.Parallel3dAxisGeometry(
    apart, dpart)
    
projector1 = AttenuatedRayTransform(
    domain, geometry, impl='niftyrec')
projector2 = RayTransform(
    domain, geometry, impl='astra_cuda')
    
data = odl.util.shepp_logan(domain)

assert projector1(data) == projector2(data)
```

Perhaps we should add a slow test that verifies that this is the case? You could add `AttenuatedRayTransform` with the right parameters [here](https://github.com/odlgroup/odl/blob/3543720d91841c602bfebf94bdbe26986a6f3718/test/largescale/tomo/ray_transform_slow_test.py#L37-L56).

niinimaki commented:
Added a helper function `spect_clintec_geometries_from_file` into `clintec_spect.py` ([57f7ffb)](https://github.com/odlgroup/odl/pull/391/commits/57f7ffb7cd1ea35ac33db30f70d8219c71da1c41).

Also fixed failing tests of earlier commit  [(4249974)](https://github.com/odlgroup/odl/pull/391/commits/4249974258c6a04c6f2997d5a935fd2a088efec2).

adler-j commented on examples/tomo/niftyrec_reconstruct.py:
```diff
@@ -0,0 +1,68 @@
+
+import numpy as np
+import odl
+from odl.tomo.operators.spect_trafo import AttenuatedRayTransform
```
from odl.tomo import AttenuatedRayTransform

should suffice

adler-j commented on examples/tomo/niftyrec_reconstruct.py:
```diff
@@ -0,0 +1,68 @@
+det_nx_mm = 2
+det_radius = 200
+n_proj = 64
+det_param = det_nx_mm * det_nx_pix / 2
```
Is there a reason for this way of doing the input? In the other examples we give the extent directly.

niinimaki replied:
No strong reason, I'll change this.

adler-j commented on examples/tomo/niftyrec_reconstruct.py:
```diff
@@ -0,0 +1,68 @@
+# Create phantoms
+vol = odl.util.phantom.derenzo_sources(domain)
+vol *= 100
+attenuation = odl.util.phantom.derenzo_sources(domain)
```
Here we could use a `derenzo_attenuation`

adler-j commented on examples/tomo/niftyrec_reconstruct.py:
```diff
@@ -0,0 +1,68 @@
+data = projector(vol)
+
+# Calculate operator norm for landweber
+op_norm_est_squared = projector.adjoint(data).norm() / vol.norm()
```
its more clear to use `power_method_opnorm`

niinimaki replied:
OK, I'll change it, this was actually a [copy](https://github.com/odlgroup/odl/blob/master/examples/tomo/stir_reconstruct.py#L44) from an older example :)

niinimaki commented:
Modified the example according to @adler-j  review.
Will change `projector.adjoint(data).norm() / vol.norm()`
to `odl.power_method_opnorm` after [this](https://github.com/odlgroup/odl/pull/495/commits/94d8c9f863e3c8c0cd2ecffbc6774db49831047f) is merged

niinimaki commented:
Added an example of how reconstruction could be performed with CLINTEC data. This example calls for the helper functions in `clintec_spect.py`

Also have added checks for `niftyrec_gpu`.

niinimaki commented:
Added `spect_clintec_geometries` into `__all__` in `clintec_data` since it was missing from the list.

adler-j commented:
I guess we should make an effort here and perhaps get this in?
